anchore_engine/services/policy_engine/engine/vulns/stores.py

```diff
@@ -1,57 +1,108 @@
 from anchore_engine.clients.services import internal_client_for, catalog
 from anchore_engine.db import (
     Image,
     get_thread_scoped_session as get_session,
-    CachedVulnerabilities,
+    ImageVulnerabilitiesReport as DbImageVulnerabilities,
 )
 from anchore_engine.services.policy_engine.api.models import ImageVulnerabilitiesReport
 from anchore_engine.subsys import logger as log
-from anchore_engine.clients.grype_wrapper import GrypeWrapperSingleton
 
-# Disabled by default, can be set in config file. Seconds for connection to cache
-DEFAULT_CACHE_CONN_TIMEOUT = -1
+# Disabled by default, can be set in config file. Seconds for connection to store
+DEFAULT_STORE_CONN_TIMEOUT = -1
 # Disabled by default, can be set in config file. Seconds for first byte timeout
-DEFAULT_CACHE_READ_TIMEOUT = -1
+DEFAULT_STORE_READ_TIMEOUT = -1
 
+REPORT_SAVE_RETRIES = 5  # number of save attempts
+REPORT_SAVE_WAIT = 1000  # wait between retries in milliseconds
 
-class CacheStatus(enum.Enum):
+
+class Status(enum.Enum):
     valid = "valid"
     stale = "stale"
     invalid = "invalid"
     missing = "missing"
 
 
-@dataclass
-class CacheRecord:
-    result: Dict
-    status: CacheStatus
+@dataclass(eq=True)
+class GrypeDBKey:
+    """
+    A key in the context of the image vulnerabilities store is the information that makes an entry in the store unique.
+    The key class has functions for generating an instance from the store's report_key and a ImageVulnerabilitiesReport
+    instance. It should also support a function for returning the status of the store entry
+
+    This key is based only on Grype DB details of the generated report.
 
+    To override the behaviour simply create a new key class and assign ImageVulnerabilitiesStore.__report_key_class__
+    to the new class
+    """
 
-@dataclass(eq=True)
-class GrypeMetadata:
-    version: str
-    db_version: str
+    # version: str
+    # db_version: str
     db_checksum: str
 
     @classmethod
-    def from_dict(cls, data):
-        return GrypeMetadata(
-            data.get("version"), data.get("db_version"), data.get("db_checksum")
+    def from_db(cls, report_key: Dict):
+        db_checksum = report_key.get("db_checksum") if report_key else None
+        if db_checksum:
+            return GrypeDBKey(db_checksum=db_checksum)
+        else:
+            raise ValueError(
+                "Invalid or unexpected report_key format {}".format(report_key)
+            )
+
+    @classmethod
+    def from_report(cls, report: ImageVulnerabilitiesReport):
+        # TODO  update this to checksum in the report after grype makes it available
+        db_checksum = (
+            GrypeWrapperSingleton.get_instance().get_current_grype_db_checksum()
         )
+        return GrypeDBKey(db_checksum=db_checksum)
 
     def to_dict(self):
         return self.__dict__
 
+    @classmethod
+    def get_report_status(cls, report_key: Dict):
```

Comment:
Is this really a class function of the Key class? Seems like a function that just takes the key as input but has no special knowledge of the class or state for the class. Binding these functions to the key class itself makes it harder to, for example, swap key implementations (but with a consistent interface) in the future.

Reply:
To be able to swap out the key implementation is the purpose of having the logic encapsulated like this. I made it a class method initially since it was returning an instance of the class. Will fix it

Reply:
Yeah I get that, not a big deal, but its a pattern I'm seeing that limits function re-use in cases where others may find a function useful but not need the broader class context since the function doesn't rely on it. And testing also requires initializing the class which can involve other dependencies. Not an issue here, but we should be careful with the pattern of making everything class methods since it can lead to a lot of code duplication or cross-class invocations that shouldn't actually exist.